src/org/traccar/database/CommandsManager.java

```diff
@@ -0,0 +1,136 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.database;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import org.traccar.BaseProtocol;
+import org.traccar.Context;
+import org.traccar.helper.Log;
+import org.traccar.model.Command;
+import org.traccar.model.CommandType;
+import org.traccar.model.Position;
+
+public class CommandsManager  extends ExtendedObjectManager<Command> {
+
+    private boolean fallbackToText;
+
+    public CommandsManager(DataManager dataManager) {
+        super(dataManager, Command.class);
+        fallbackToText = Context.getConfig().getBoolean("command.fallbackToSms");
+    }
+
+    public boolean checkDeviceCommand(long deviceId, long commandId) {
+        return !getAllDeviceItems(deviceId).contains(commandId);
+    }
+
+    public void sendCommand(Command command) throws Exception {
+        sendCommand(command, command.getDeviceId(), fallbackToText);
+    }
+
+    public void sendCommand(long commandId, long deviceId) throws Exception {
+        sendCommand(getById(commandId), deviceId, false);
+    }
+
+    public void sendCommand(Command command, long deviceId, boolean fallbackToText) throws Exception {
+        if (command.getTextChannel()) {
+            Position lastPosition = Context.getIdentityManager().getLastPosition(deviceId);
+            String phone = Context.getIdentityManager().getById(deviceId).getPhone();
+            if (lastPosition != null) {
+                BaseProtocol protocol = Context.getServerManager().getProtocol(lastPosition.getProtocol());
+                protocol.sendTextCommand(phone, command);
+            } else if (command.getType().equals(Command.TYPE_CUSTOM)) {
+                if (Context.getSmppManager() != null) {
+                    Context.getSmppManager().sendMessageSync(phone, command.getString(Command.KEY_DATA), true);
+                } else {
+                    throw new RuntimeException("SMPP client is not enabled");
+                }
+            } else {
+                throw new RuntimeException("Command " + command.getType() + " is not supported");
+            }
+        } else {
+            ActiveDevice activeDevice = Context.getConnectionManager().getActiveDevice(deviceId);
+            if (activeDevice != null) {
+                activeDevice.sendCommand(command);
+            } else {
+                if (fallbackToText) {
+                    command.setTextChannel(true);
+                    sendCommand(command, deviceId, false);
+                } else {
+                    throw new RuntimeException("Device is not online");
+                }
+            }
+        }
+    }
+
+    public Collection<Long> getSupportedCommands(long deviceId) {
```

Comment:
Now it returns linked to device (with groups) supported by device protocol according to status (online/offline) Saved Commands. Text if it is offline, text and data if it is online